File: src/main/java/raylras/zen/langserver/provider/DefinitionProvider.java
package raylras.zen.langserver.provider;

import org.antlr.v4.runtime.tree.ParseTree;
import org.eclipse.lsp4j.DefinitionParams;
import org.eclipse.lsp4j.Location;
import org.eclipse.lsp4j.LocationLink;
import org.eclipse.lsp4j.jsonrpc.messages.Either;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import raylras.zen.code.resolve.SymbolResolver;
import raylras.zen.code.symbol.ParseTreeLocatable;
import raylras.zen.code.symbol.Symbol;
import raylras.zen.langserver.Document;
import raylras.zen.util.CSTNodes;
import raylras.zen.util.Position;
import raylras.zen.util.Range;

import java.util.Collection;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public class DefinitionProvider {

    private static final Logger logger = LoggerFactory.getLogger(DefinitionProvider.class);

    public static CompletableFuture<Either<List<? extends Location>, List<? extends LocationLink>>> definition(Document doc, DefinitionParams params) {
        return doc.getUnit().map(unit -> CompletableFuture.supplyAsync(() -> {
            Position cursor = Position.of(params.getPosition());
            ParseTree cst = CSTNodes.getCstAtPosition(unit.getParseTree(), cursor);
            org.eclipse.lsp4j.Range originSelectionRange = Range.of(cst).toLspRange();
            Collection<Symbol> symbols = SymbolResolver.lookupSymbol(cst, unit);
            return Either.<List<? extends Location>, List<? extends LocationLink>>forRight(symbols.stream()
                    .filter(symbol -> symbol instanceof ParseTreeLocatable)
                    .map(symbol -> toLocationLink(symbol, originSelectionRange))
                    .toList());
        })).orElseGet(DefinitionProvider::empty);
    }

    public static CompletableFuture<Either<List<? extends Location>, List<? extends LocationLink>>> empty() {
        return CompletableFuture.completedFuture(null);
    }

    private static LocationLink toLocationLink(Symbol symbol, org.eclipse.lsp4j.Range originSelectionRange) {
        ParseTreeLocatable locatable = ((ParseTreeLocatable) symbol);
        String uri = locatable.getUri();
        org.eclipse.lsp4j.Range range = locatable.getRange().toLspRange();
        org.eclipse.lsp4j.Range selectionRange = locatable.getSelectionRange().toLspRange();
        return new LocationLink(uri, range, selectionRange, originSelectionRange);
    }

}

File: language-server/src/main/kotlin/raylras/intellizen/languageserver/provider/DefinitionProvider.kt
package raylras.intellizen.languageserver.provider

import org.antlr.v4.runtime.tree.RuleNode
import org.antlr.v4.runtime.tree.TerminalNode
import org.eclipse.lsp4j.DefinitionParams
import org.eclipse.lsp4j.LocationLink
import raylras.intellizen.CompilationUnit
import raylras.intellizen.Visitor
import raylras.intellizen.languageserver.util.toLspRange
import raylras.intellizen.languageserver.util.toTextPosition
import raylras.intellizen.parser.ZenScriptParser.*
import raylras.intellizen.resolve.resolveSemantics
import raylras.intellizen.symbol.ParseTreeLocatable
import raylras.intellizen.util.TextRange
import raylras.intellizen.util.contains
import raylras.intellizen.util.getTerminalAt
import raylras.intellizen.util.textRange

object DefinitionProvider {
    fun definition(unit: CompilationUnit, params: DefinitionParams): List<LocationLink> {
        val cursor = params.position.toTextPosition()
        val terminal = unit.parseTree.getTerminalAt(cursor) ?: return emptyList()
        val visitor = DefinitionVisitor(unit, terminal)
        unit.accept(visitor)
        return visitor.result.toList()
    }
}

private class DefinitionVisitor(private val unit: CompilationUnit, private val terminal: TerminalNode) : Visitor<Unit>() {
    var result = emptySequence<LocationLink>()

    override fun visitQualifiedName(ctx: QualifiedNameContext) {
        result = resolveSemantics(ctx, unit)
            .filterIsInstance<ParseTreeLocatable>()
            .map { it.toLocationLink(ctx.textRange) }
    }

    override fun visitSimpleNameExpr(ctx: SimpleNameExprContext) {
        result = resolveSemantics(ctx, unit)
            .filterIsInstance<ParseTreeLocatable>()
            .map { it.toLocationLink(ctx.textRange) }
    }

    override fun visitMemberAccessExpr(ctx: MemberAccessExprContext) {
        when (terminal) {
            in ctx.simpleName() -> {
                result = resolveSemantics(ctx, unit)
                    .filterIsInstance<ParseTreeLocatable>()
                    .map { it.toLocationLink(ctx.simpleName().textRange) }
            }

            else -> {
                visitChildren(ctx)
            }
        }
    }

    override fun visitChildren(node: RuleNode) {
        for (i in 0 until node.childCount) {
            val child = node.getChild(i)
            if (terminal in child) {
                return child.accept(this)
            }
        }
    }
}

private fun ParseTreeLocatable.toLocationLink(originSelectionRange: TextRange): LocationLink {
    val uri = this.path.toUri().toString()
    val range = this.textRange.toLspRange()
    val selectionRange = this.simpleNameTextRange.toLspRange()
    return LocationLink(uri, range, selectionRange, originSelectionRange.toLspRange())
}
